streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -131,10 +131,18 @@ public void setGlobalProcessorContext(final InternalProcessorContext globalProce
         }
 
         final List<StateStore> stateStores = topology.globalStateStores();
+        final Map<String, String> storeNameToTopic = topology.storeToChangelogTopic();
+        final Set<String> stateStoreTopics = new HashSet<>();
         for (final StateStore stateStore : stateStores) {
             globalStoreNames.add(stateStore.name());
+            final String sourceTopic = storeNameToTopic.get(stateStore.name());
+            stateStoreTopics.add(sourceTopic);
             stateStore.init(globalProcessorContext, stateStore);
         }
+
+        // prune topic-partitions not associated with any global state store from checkpointFileCache
+        checkpointFileCache.keySet().removeIf(e -> !stateStoreTopics.contains(e.topic()));
```

Comment:
Thinking about the description in the Jira once more, the issue is a mix of a bug and incorrect usage of the API. If you change the topology (ie, change the topic name of the global store), it is actually required (ie, the users responsibility) to reset the application: this reset would delete the local state directory including the checkpoint file and avoid the issue.

Hence, I am wondering, if we should better throw an exception if we detect a corrupted checkpoint file (ie, a checkpoint file that does not match what we expect to find in it), with a clear error message that the user needs to reset the application? If we argue this way, the bug is that we did not check if the checkpoint file is corrupted and fail with an unclear NPE.

As an alternative, we should also be a little bit more resilient against a corrupted checkpoint file and follow the proposed approach to clean the `checkpointFileCache`. For this case, we should at least log a WARNING that the checkpoint file was corrupted.

Personally, I tend to prefer throwing an exception, because otherwise an unintended change of the topic name may be undetected (if a user does not carefully monitor the logs), and hence execute incorrect code. Throwing a descriptive exception would guard against this case, because we force users to explicitly delete the corrupted checkpoint file.

 \cc @guozhangwang @bbejeck @vvcephei @ableegoldman @cadonna @abbccdda

Reply:
Thanks for the feedback, @mjsax. I was not aware that a simple GlobalKTable state store name change required an application-reset -- is this documented somewhere? Also, I was under the impression that the .checkpoint file was an implementation detail that is maintained automagically. Thus, I thought it'd be reasonable for the global task to make changes as needed (although, I do think logging a WARNING each time a topic is removed is a good idea).  In any case, I'll defer to the experts here and will make the appropriate changes.

Reply:
Not sure if the documentation is very explicit about it. However, a `GlobalKTable` is persistent to disk by default (similar to other persistent stores). If you run your program using topic "A" for the GlobalKTable, some data from topic "A" will be put into the table, ie, persistent to local disk. If you change the topic to "B", you intent to load different data, thus, you _need_ to delete the local state -- otherwise, the data from topic "A" would still be in the global store. On startup, the local persistent store is loaded first and afterwards data from the topic would be loaded. Hence, you may end up with data from topic "A" and topic "B" in your global store.

Each time you change _anything_, you should change the `application.id` or reset the application (there are exception when a reset is not required, but to know _when_ an exception applies is tricky -- hence it's recommended to just reset). Otherwise, old state may "carry over" corrupting the application on restart.

Btw: if you change the store name, too, (not just the topic name, as described in the Jiar) data corruption should actually not be a problem and it should not happen that you carry over data from the old store to the new store, because it would be two different directories, and the store for topic "B" would be empty. However, the `.checkpoint` file is used for all stores though, and would still be corrupted I guess, and the NPE may happen also if you change both, topic and store name. Also, if you change the store name, you should still reset to clean up the state directory and free up disk space of the old store.

> Also, I was under the impression that the .checkpoint file was an implementation detail that is maintained automagically.

That is correct, however, a corrupted `.checkpoint` file is an indicator of what I described above, ie, potential data corruption by carrying over old state, or wasting disk space. Hence, I am not sure if we should "cleanup" the checkpoint file without telling the user about this potential issue?

Maybe we can improve the documentation as part of this PR, too?

Reply:
I see. Given the fact that:

> Hence, you may end up with data from topic "A" and topic "B" in your global store.

isn't the only possible solution to throw an exception with and require manual intervention? i.e. if cleaning up the corrupted `.checkpoint` file can still result in data corruption, we cannot possibly continue, correct?